File: agent/optimizer.py
"""Post-processing: DAG validation, critical path, load balance check.

Pure-Python — no LLM. Operates on the project dict produced by `decomposer`.
"""
from __future__ import annotations

from typing import Any


def all_stories(project: dict[str, Any]) -> list[dict[str, Any]]:
    return [s for epic in project.get("epics", []) for s in epic.get("stories", [])]


def validate_dag(project: dict[str, Any]) -> dict[str, Any]:
    """Check that every depends_on refers to a real story and there are no cycles.

    Returns a report dict with `unknown_refs`, `cycles`, and `ok`.
    """
    stories = all_stories(project)
    by_id = {s["id"]: s for s in stories}

    unknown_refs: list[tuple[str, str]] = []
    for s in stories:
        for dep in s.get("depends_on", []) or []:
            if dep not in by_id:
                unknown_refs.append((s["id"], dep))

    # Cycle detection via DFS
    WHITE, GRAY, BLACK = 0, 1, 2
    color = {sid: WHITE for sid in by_id}
    cycles: list[list[str]] = []

    def dfs(sid: str, stack: list[str]) -> None:
        color[sid] = GRAY
        stack.append(sid)
        for dep in by_id[sid].get("depends_on", []) or []:
            if dep not in by_id:
                continue
            if color[dep] == GRAY:
                # cycle: extract from stack
                if dep in stack:
                    cycles.append(stack[stack.index(dep) :] + [dep])
            elif color[dep] == WHITE:
                dfs(dep, stack)
        stack.pop()
        color[sid] = BLACK

    for sid in by_id:
        if color[sid] == WHITE:
            dfs(sid, [])

    return {
        "unknown_refs": unknown_refs,
        "cycles": cycles,
        "ok": not unknown_refs and not cycles,
    }


def critical_path(project: dict[str, Any]) -> list[str]:
    """Longest dependency chain by estimate_points."""
    stories = all_stories(project)
    by_id = {s["id"]: s for s in stories}

    memo: dict[str, tuple[float, list[str]]] = {}

    def longest(sid: str) -> tuple[float, list[str]]:
        if sid in memo:
            return memo[sid]
        s = by_id.get(sid)
        if s is None:
            memo[sid] = (0, [])
            return memo[sid]
        best_cost = 0.0
        best_path: list[str] = []
        for dep in s.get("depends_on", []) or []:
            cost, path = longest(dep)
            if cost > best_cost:
                best_cost = cost
                best_path = path
        pts = float(s.get("estimate_points") or 0)
        memo[sid] = (best_cost + pts, [*best_path, sid])
        return memo[sid]

    if not by_id:
        return []
    best = max((longest(sid) for sid in by_id), key=lambda x: x[0])
    return best[1]


def load_by_member(project: dict[str, Any]) -> dict[str, float]:
    totals: dict[str, float] = {}
    for s in all_stories(project):
        a = s.get("assignee") or "Unassigned"
        totals[a] = totals.get(a, 0.0) + float(s.get("estimate_points") or 0)
    return totals


def annotate(project: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of project with execution_plan.critical_path filled in if missing."""
    plan = project.setdefault("execution_plan", {})
    if not plan.get("critical_path"):
        plan["critical_path"] = critical_path(project)
    plan["_load_by_member"] = load_by_member(project)
    plan["_dag_report"] = validate_dag(project)
    return project

File: web/lib/optimizer.ts
// Port of agent/optimizer.py — pure logic, no LLM.
import type { Project, Story } from "./schema";

export function allStories(project: Project): Story[] {
  return project.epics.flatMap((e) => e.stories);
}

export type DagReport = {
  unknown_refs: [string, string][];
  cycles: string[][];
  ok: boolean;
};

export function validateDag(project: Project): DagReport {
  const stories = allStories(project);
  const byId = new Map(stories.map((s) => [s.id, s]));

  const unknown: [string, string][] = [];
  for (const s of stories) {
    for (const dep of s.depends_on || []) {
      if (!byId.has(dep)) unknown.push([s.id, dep]);
    }
  }

  const WHITE = 0,
    GRAY = 1,
    BLACK = 2;
  const color = new Map<string, number>();
  for (const id of byId.keys()) color.set(id, WHITE);
  const cycles: string[][] = [];

  function dfs(sid: string, stack: string[]) {
    color.set(sid, GRAY);
    stack.push(sid);
    for (const dep of byId.get(sid)?.depends_on || []) {
      if (!byId.has(dep)) continue;
      const c = color.get(dep);
      if (c === GRAY) {
        const idx = stack.indexOf(dep);
        if (idx !== -1) cycles.push([...stack.slice(idx), dep]);
      } else if (c === WHITE) {
        dfs(dep, stack);
      }
    }
    stack.pop();
    color.set(sid, BLACK);
  }

  for (const id of byId.keys()) {
    if (color.get(id) === WHITE) dfs(id, []);
  }

  return {
    unknown_refs: unknown,
    cycles,
    ok: unknown.length === 0 && cycles.length === 0,
  };
}

export function criticalPath(project: Project): string[] {
  const stories = allStories(project);
  const byId = new Map(stories.map((s) => [s.id, s]));
  const memo = new Map<string, { cost: number; path: string[] }>();

  function longest(sid: string): { cost: number; path: string[] } {
    if (memo.has(sid)) return memo.get(sid)!;
    const s = byId.get(sid);
    if (!s) {
      const r = { cost: 0, path: [] as string[] };
      memo.set(sid, r);
      return r;
    }
    let bestCost = 0;
    let bestPath: string[] = [];
    for (const dep of s.depends_on || []) {
      const r = longest(dep);
      if (r.cost > bestCost) {
        bestCost = r.cost;
        bestPath = r.path;
      }
    }
    const pts = Number(s.estimate_points || 0);
    const result = { cost: bestCost + pts, path: [...bestPath, sid] };
    memo.set(sid, result);
    return result;
  }

  if (byId.size === 0) return [];
  let best = { cost: -1, path: [] as string[] };
  for (const id of byId.keys()) {
    const r = longest(id);
    if (r.cost > best.cost) best = r;
  }
  return best.path;
}

export function loadByMember(project: Project): Record<string, number> {
  const totals: Record<string, number> = {};
  for (const s of allStories(project)) {
    const a = s.assignee || "Unassigned";
    totals[a] = (totals[a] || 0) + Number(s.estimate_points || 0);
  }
  return totals;
}

/** Return a copy of `project` with execution_plan annotations filled in. */
export function annotate(project: Project): Project {
  const next: Project = JSON.parse(JSON.stringify(project));
  const plan = next.execution_plan;
  if (!plan.critical_path || plan.critical_path.length === 0) {
    plan.critical_path = criticalPath(next);
  }
  plan._load_by_member = loadByMember(next);
  plan._dag_report = validateDag(next);
  return next;
}
